doc/specs/Proto extensions-spec.md

```diff
@@ -0,0 +1,139 @@
+for the full version see the spec for cascading default + user settings):
+
+1. Generate profiles from the defaults file
+2. Generate profiles from the dynamic profile generators
+3. Layer the user settings on top of all the profiles created in steps 1 and 2
+4. Validate the settings
+
+To allow for installations to add in their snippets of json, I propose the addition of a new step
+in between 2 and 3:
+
+1. Generate profiles from the defaults file
+2. Generate profiles from the dynamic profile generators
+3. **Incorporate the additional provided json stubs** - these stubs could be:
+   1. modifications to existing profiles
+   2. full profiles
+   3. colour schemes/modifications to colour schemes
+4. Layer the user settings on top of all the profiles created in steps 1 through 3
+5. Validate the settings
+
+### Specifications of the json stubs
+
+As written above, the json stubs could be modifications to existing profiles, full profiles, or colour schemes.
+
+#### Modifications to existing profiles
+
+For modifications to existing profiles, the json stub would need to indicate which profile it wishes to modify.
+A straightforward way to do this would be by providing the name of the profile. Since this modification would
+occur at the step _before_ any user changes to names are layered, the names of the profiles are consistent and
+can be used for identification. For example, the name of a profile generated by the WSL profile generator is
+simply the argument to "wsl.exe -d \<name\>", the name of a profile generated by the Powershell profile generator
+would be "Powershell Core" or "Powershell (ARM)" etc depending on the version.
+
+We might run into the case where multiple json stubs modify the same profile and so they override each other. We
+will need some sort of hierarchy to determine which stub to ultimately use. My initial thought is that we could
+check the owner of the json file and see if it matches the profile they are trying to modify (so something like
+only Powershell can modify the Powershell profiles and only Ubuntu can modify the Ubuntu profile etc). Independent
+parties will need to create full profiles with the desired commandline argument instead.
+
+#### Full profile stubs
+
+Technically, full profile stubs do not need to contain anything (they could just be '\{\}'). However we should
+have some qualifying minimum criteria before we accept a stub as a full profile. I suggest that we only create
+new profile objects from stubs that contain at least the following
+
+* A name (and this name _should not_ match the names of existing profiles as described above, because that would cause us to interpret this profile as modifying an existing one)
+* A commandline argument
+* A unique GUID - this is so we avoid conflicts if several different creators name their profile the same thing
+
+As in the case of the dynamic profile generator, if we create a profile that did not exist before (i.e. does not

```

Comment:
What does this stamped stub look like in settings.json? Will it have a "source" to denote that it came from this stub? Is it just the stub copied/pasted directly? Or will it look small/lightweight like the dynamically generated profiles?

Reply:
It will be copied/pasted directly (just as if the user typed it up themselves), which would allow the user to customize it further as well. We might add a source to it too - pending discussion on that.

Reply:
It will be lightweight like the dynamically generated profiles and it will have a source!